debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSignalBasedIncrementalSnapshotChangeEventSource.java

```diff
@@ -57,4 +68,155 @@ protected Table refreshTableSchema(Table table) throws SQLException {
+                }
+            }
+        }
+        return statement;
+    }
+
+    private String buildUnionBasedChunkQuery(Table table, int limit, Optional<String> additionalCondition) {
+        List<String> boundConditions = Collections.emptyList();
+
+        // Add condition when this is not the first query
+        if (context.isNonInitialChunk()) {
+            // Window boundaries
+            boundConditions = getBoundConditions(table);
+        }
+
+        List<String> orderByColumnNames = getQueryColumns(table).stream()
+                .map(Column::name)
+                .collect(Collectors.toList());
+
+        return jdbcConnection.buildUnionBasedSelectWithRowLimits(table.id(),
+                limit,
+                buildProjection(table),
+                boundConditions,
+                additionalCondition,
+                orderByColumnNames);
+    }
+
+    private List<String> getBoundConditions(Table table) {
+        // Provides boundary conditions for the incremental snapshot query
+        // For one column the condition will be (? will always be the last value seen for the given column)
+        // (k1 > ?) AND (k1 <= ?)
+        // For two columns
+        // (k1 > ?) AND (k1 <= ?)
+        // (k1 = ? AND k2 > ?) AND (k1 <> ? OR k2 <= ?)
+        // For three columns
+        // (k1 > ?) AND (k1 <= ?)
+        // (k1 = ? AND k2 > ?) AND (k1 <> ? OR k2 <= ?)
+        // (k1 >= ? AND k2 >= ? AND k3 > ?) AND (k1 <> ? OR k2 <> ? OR k3 <= ?)
+        // For four columns
+        // (k1 > ?) AND (k1 <= ?)
+        // (k1 = ? AND k2 > ?) AND (k1 <> ? OR k2 <= ?)
+        // (k1 >= ? AND k2 >= ? AND k3 > ?) AND (k1 <> ? OR k2 <> ? OR k3 <= ?)
+        // (k1 = ? AND k2 = ? AND k3 = ? AND k4 > ?) AND (k1 <> ? OR k2 <> ? OR k3 <> ? OR k4 <= ?)
+        // etc.
+        List<String> conditionList = new ArrayList<>();
+        final List<Column> pkColumns = getQueryColumns(table);
+
+        for (int i = 0; i < pkColumns.size(); i++) {
+            StringBuilder sql = new StringBuilder("(");
+
+            for (int j = 0; j < i + 1; j++) {
+                final boolean isLastIterationForJ = (i == j);
+
+                sql.append(jdbcConnection.quotedColumnIdString(pkColumns.get(j).name()));
+                sql.append(isLastIterationForJ ? " > ?" : " = ?");
+
+                if (!isLastIterationForJ) {
+                    sql.append(" AND ");
+                }
+            }
+
+            sql.append(") AND (");
+
+            for (int j = 0; j < i + 1; j++) {
+                final boolean isLastIterationForJ = (i == j);
+
+                sql.append(jdbcConnection.quotedColumnIdString(pkColumns.get(j).name()));
+                sql.append(isLastIterationForJ ? " <= ?" : " <> ?");
+
+                if (!isLastIterationForJ) {
+                    sql.append(" OR ");
+                }
+            }
+
+            sql.append(")");
+            conditionList.add(sql.toString());
+        }
+
+        return conditionList;
```

Comment:
to avoid messing up with the indexes and improve the readability a little. can we instead do something like this
```suggestion
        final List<Column> pkColumns = getQueryColumns(table);
        return IntStream.rangeClosed(1, pkColumns.size())
            .mapToObj(i -> pkColumns.subList(0, i))
            .map(columns -> {
                var firstConditionForSubquery ="(" +  Strings.join(" = ? AND ", columns, column -> jdbcConnection.quotedColumnIdString(column.name())) + " > ?)";
                var secondConditionForSubQuery ="(" +  Strings.join(" <> ? OR ", columns, column -> jdbcConnection.quotedColumnIdString(column.name())) + " <= ?)";
                return firstConditionForSubquery + " AND " + secondConditionForSubQuery;
            }).collect(Collectors.toList());
```

Reply:
I wanted to use the same notation as the default query builder is implemented for easier comparison between them.

Reply:
Should not we compare the results of what's generated instead of comparing the source code? I think keeping indexing like this and checking for last iteration in internal loop just brings additional complexity to otherwise quite simple things. Since original implementation does a little different thing there it looks better. 

So my suggestion would be to have some(or even maybe one) tests(maybe parameterized) to just showcase how the difference looks like without copy-pasting the tests from another class because here we don't need to cover all the corner cases of the original implementation.
And then proceed with the refactoring of this part. 
Not sure however if we have time to address this right now and if we don't we can merge without it.